torch/utils/tensorboard/_pytorch_graph.py

```diff
@@ -0,0 +1,267 @@
+import numpy as np
+import time
+from collections import OrderedDict
+
+from tensorboard.compat.proto.config_pb2 import RunMetadata
+from tensorboard.compat.proto.graph_pb2 import GraphDef
+from tensorboard.compat.proto.step_stats_pb2 import StepStats, DeviceStepStats, NodeExecStats, AllocatorMemoryUsed
+from tensorboard.compat.proto.versions_pb2 import VersionDef
+
+import torch
+from ._proto_graph import node_proto
+from torch.onnx.utils import OperatorExportTypes
+
+
+methods_OP = ['attributeNames', 'hasMultipleOutputs', 'hasUses', 'inputs',
+              'kind', 'outputs', 'outputsSize', 'scopeName']
+methods_IO = ['node', 'offset', 'uniqueName']  # 'unique' <int> , 'type' <Tensor<class 'torch._C.Type'>>
+
+
+class NodeBase(object):
+    def __init__(self, uniqueName=None, inputs=None, scope=None, tensor_size=None, op_type='UnSpecified', attributes=''):
+        self.uniqueName = uniqueName
+        self.inputs = inputs
+        self.tensor_size = tensor_size
+        self.kind = op_type
+        self.attributes = attributes
+        if scope is not None:
+            self.scope = scope
+
+    def __repr__(self):
+        repr = []
+        repr.append(str(type(self)))
+        for m in dir(self):
+            if '__' not in m:
+                repr.append(m + ': ' + str(getattr(self, m)) + str(type(getattr(self, m))))
+        return '\n'.join(repr) + '\n\n'
+
+
+class NodePy(NodeBase):
+    def __init__(self, NodeCpp, valid_methods):
+        super(NodePy, self).__init__(NodeCpp)
+        self.valid_methods = valid_methods[:]
+        self.inputs = []
+
+        for m in self.valid_methods:
+            if m == 'inputs' or m == 'outputs':
+                list_of_node = list(getattr(NodeCpp, m)())
+                io_unique_names = []
+                io_tensor_sizes = []
+                for n in list_of_node:
+                    io_unique_names.append(n.uniqueName())
+                    if n.type().kind() == 'CompleteTensorType':
+                        io_tensor_sizes.append(n.type().sizes())
+                    else:
+                        io_tensor_sizes.append(None)
+
+                setattr(self, m, io_unique_names)
+                setattr(self, m + 'tensor_size', io_tensor_sizes)
+
+            else:
+                setattr(self, m, getattr(NodeCpp, m)())
+
+
+class NodePyIO(NodePy):
+    def __init__(self, NodeCpp, input_or_output=None):
+        super(NodePyIO, self).__init__(NodeCpp, methods_IO)
+        try:
+            tensor_size = NodeCpp.type().sizes()
+        except RuntimeError:
+            tensor_size = [1, ]  # fail when constant model is used.
+        self.tensor_size = tensor_size
+        self.kind = 'Parameter'
+        if input_or_output:
+            self.input_or_output = input_or_output
+            self.kind = 'IO Node'
```

Comment:
Is the `kind` here selected out of a fixed pool of valid names, or is it just purely descriptive?

Reply:
@lanpa can you comment here? Thanks.

Reply:
purely descriptive. It will be shown in the detailed information pop-up in tensorboard. nodes of type `Node_py_OP` will get its `kind` by calling the `.kind()`

Reply:
Thanks! Created task to add comment text.